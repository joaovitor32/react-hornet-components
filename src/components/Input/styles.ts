import styled, { css, SimpleInterpolation } from 'styled-components';

import { Palette, Dimension } from '../../types';

interface DivsxProps {
  palette: Palette;
  dimensions: Dimension;
  sx?: SimpleInterpolation;
}

const Container = styled.div<DivsxProps>`
  position: relative;

  ${(props) =>
    css`
      width: ${props.dimensions.width};
      height: ${props.dimensions.height};
    `};

  label {
    position: absolute;
    font-family: inherit;
    font-size: 1rem;
    font-weight: 400;
    line-height: inherit;
    left: 8px;
    top: 12px;
    padding: 0px 4px 0px 0px;
    transition: all 0.3s ease;

    ${(props) =>
      css`
        color: ${props.palette.primary};
        background: ${props.palette.tertiary};
      `};
  }

  input {
    outline: none;
    position: absolute;
    font-family: inherit;
    font-size: 16px;
    font-weight: 400;
    line-height: inherit;
    top: 0;
    left: 0;
    z-index: 1;
    resize: none;
    width: 100%;
    height: auto;
    padding: 12px 16px;
    border-radius: 4px;
    background: transparent;
    transition: all 0.2s ease-in-out;

    ${(props) =>
      css`
        border: 1px solid ${props.palette.primary};
        color: ${props.palette.primary};
      `};

    &::placeholder {
      opacity: 0;
      visibility: hidden;
      color: transparent;
    }

    &:focus {
      outline: none;

      ${(props) =>
        css`
          border: 1px solid ${props.palette.secondary};
          color: ${props.palette.secondary};
        `};

      & ~ label {
        padding-left: 4px;
        top: -6px;
        left: 8px;
        z-index: 5;
        font-size: 12px;
        font-weight: 500;
        transition: all 0.3s ease-in-out;

        ${(props) =>
          css`
            color: ${props.palette.secondary};
          `};
      }
    }

    &:not(:placeholder-shown) {
      outline: none;
      &:not(:focus) {
        & ~ label {
          padding-left: 4px;
          left: 8px;
          top: -6px;
          z-index: 12;
          font-size: 12px;
          font-weight: 300;
          transition: all 0.3s ease-in-out;
        }
      }
    }
  }

  ${({ sx }) =>
    !!sx &&
    css`
      ${sx}
    `}
`;

export { Container };
