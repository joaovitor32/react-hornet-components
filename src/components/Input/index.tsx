import React from 'react';

import { Palette, Dimension, InputProps } from '../../types';

import { Container } from './styles';

const Input: React.FC<InputProps<Palette, Dimension>> = ({ label, name, palette, dimensions, sx, ...rest }) => (
  <Container dimensions={dimensions} palette={palette} sx={sx}>
    <input name={name} {...rest} />
    <label htmlFor={name}>{label}</label>
  </Container>
);

export default Input;
