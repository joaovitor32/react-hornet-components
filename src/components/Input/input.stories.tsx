import React from 'react';

import { Meta, Story } from '@storybook/react';

import Input from '.';

export default {
  title: 'Components/Input',
  component: Input,
} as Meta;

const Template: Story<any> = (args) => <Input {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  palette: {
    primary: 'green',
    secondary: 'purple',
    tertiary: 'white',
  },
  dimensions: {
    height: '24px',
    width: '240px',
  },
  label: 'texto input',
  name: 'texto input',
  placeholder: 'none',
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => console.log(e),
};
