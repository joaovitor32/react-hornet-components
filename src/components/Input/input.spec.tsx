import React from 'react';

import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import Input from '.';

describe('<Input />', () => {
  it('should render correctly, --success case', () => {
    const { container } = render(
      <Input
        title=""
        label="name"
        dimensions={{ width: '', height: '' }}
        palette={{ primary: '', secondary: '', tertiary: '' }}
      />
    );

    expect(container.querySelector('input')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
