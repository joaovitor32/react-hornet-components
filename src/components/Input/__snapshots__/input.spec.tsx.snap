// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render correctly, --success case 1`] = `
<div
  class="sc-bcXHqe jFassX"
>
  <input
    title=""
  />
  <label>
    name
  </label>
</div>
`;
